kratos/linear_solvers/monotonicity_preserving_solver.h

```diff
@@ -0,0 +1,386 @@
+    }
+
+    /** Some solvers may require a minimum degree of knowledge of the structure of the matrix. To make an example
+     * when solving a mixed u-p problem, it is important to identify the row associated to v and p.
+     * another example is the automatic prescription of rotation null-space for smoothed-aggregation solvers
+     * which require knowledge on the spatial position of the nodes associated to a given dof.
+     * This function is the place to eventually provide such data
+     */
+    void ProvideAdditionalData(
+        SparseMatrixType& rA,
+        VectorType& rX,
+        VectorType& rB,
+        typename ModelPart::DofsArrayType& rdof_set,
+        ModelPart& r_model_part
+    ) override
+    {
+        mpLinearSolver->ProvideAdditionalData(rA,rX,rB,rdof_set,r_model_part);
+    }
+
+    void InitializeSolutionStep (SparseMatrixType& rA, VectorType& rX, VectorType& rB) override
+    {
+        mpLinearSolver->InitializeSolutionStep(rA,rX,rB);
+    }
+
+    /** This function is designed to be called at the end of the solve step.
+     * for example this is the place to remove any data that we do not want to save for later
+    @param rA. System matrix
+    @param rX. Solution vector. it's also the initial guess for iterative linear solvers.
+    @param rB. Right hand side vector.
+    */
+    void FinalizeSolutionStep (SparseMatrixType& rA, VectorType& rX, VectorType& rB) override
+    {
+        mpLinearSolver->FinalizeSolutionStep(rA,rX,rB);
+    }
+
+    /** This function is designed to clean up all internal data in the solver.
+     * Clear is designed to leave the solver object as if newly created.
+     * After a clear a new Initialize is needed
+     */
+    void Clear() override
+    {
+        mpLinearSolver->Clear();
+    }
+
+    /**
+    @param rA. System matrix
+    @param rX. Solution vector. it's also the initial
+    guess for iterative linear solvers.
+    @param rB. Right hand side vector.
+    */
+    bool Solve(SparseMatrixType& rA, VectorType& rX, VectorType& rB) override
+    {
+        if(this->IsNotConsistent(rA, rX, rB))
+            return false;
+
+        double *values_vector = rA.value_data().begin();
+        std::size_t *index1_vector = rA.index1_data().begin();
+        std::size_t *index2_vector = rA.index2_data().begin();
+
+
+        IndexPartition<std::size_t>(rA.size1()).for_each(
+            [&](std::size_t i)
+            {
+                for (std::size_t k = index1_vector[i]; k < index1_vector[i + 1]; k++) {
+                    const double value = values_vector[k];
+                    if (value > 0.0) {
+                        const auto j = index2_vector[k];
+                        if (j > i) {
+                            rA(i,i) += value;
+                            rA(i,j) -= value;
+                            rA(j,i) -= value;
+                            rA(j,j) += value;
+                            rB[i] += value*rX[j] - value*rX[i];
```

Comment:
I just found this is wrong because here we do not need rX[j] but the actual nodal value. Maybe we can move this whole trick to ProvideAdditionalData where model part is given as input. What do you think? @RiccardoRossi @philbucher

Reply:
ouch, the be honest i don t think we should do this.

the assumption in my opinion is thatbwe start from a sokution that is good, and we compite a correction that does not spoil it...

Reply:
ouch, the be honest i don t think we should do this.

the assumption in my opinion is thatbwe start from a sokution that is good, and we compite a correction that does not spoil it...

Reply:
I am not sure if we are talking about the same thing. What I mean is that to "assemble" rhs (-=prod(LHS, X) it should use nodal values, not rX.

Reply:
i got you, but i believe that nodal value cannot be accessed (read, shouldn t ) from the function.

this function can u nly modify the output, that is the x you solve for...maybe we d better speak about this on voice

Reply:
After discussing this with @RiccardoRossi we agreed that we need to access the actual nodal value. The rationale could be like following:

The system is Ax = b, but we do not solve this, we solve Adx = b - Ax.

If we use instead A* = A + B to ensure that A* does not have any non-diagonal term greater than zero, then what we will solve is:

`(A+B)dx = b - Ax - Bx`